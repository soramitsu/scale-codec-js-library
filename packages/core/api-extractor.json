{
  "extends": "../../api-extractor.json",
  "mainEntryPointFilePath": "<projectFolder>/types/core/src/lib.d.ts",
  "dtsRollup": {
    "untrimmedFilePath": "dist/lib.d.ts"
  }
}
