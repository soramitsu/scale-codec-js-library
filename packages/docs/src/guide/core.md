# Core

This library is a pure low-level performant implementation of SCALE.

## Supported types

-   **Primitive:**

    | Spec    | JS Type             | Details                                                                                                                                                             |
    | ------- | ------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
    | Int     | `number`, `bigint`  | signed & unsigned, 8/16/32/64/128/etc bits. For ints with bits less than or equal 32 both `number` & `bigint` could be used, for 64+ bits integers - only `bigint`. |
    | Compact | `number`, `bigint`  | -                                                                                                                                                                   |
    | String  | `string`            | -                                                                                                                                                                   |
    | Bool    | `boolean`           | -                                                                                                                                                                   |
    | Void    | `null`, `undefined` | `()` in Rust. Unfortunately, JavaScript doesn't have zero-cost abstractions, so this codec could be used to handle it.                                              |

-   **Higher-order:**

    | Spec   | JS Type                                        |
    | ------ | ---------------------------------------------- |
    | Array  | `Array`                                        |
    | Vector | `Array`                                        |
    | Tuple  | `Array`, but tuple in TypeScript               |
    | Set    | `Set`                                          |
    | Map    | `Map`                                          |
    | Struct | Plain Object                                   |
    | Enum   | `Enum` from [`@scale-codec/enum`](/guide/enum) |

-   **Special:**

    -   Efficient codec for arrays of bytes, or `[u8; x]` in Rust and `Uint8Array` in JS
    -   Efficient codec for vectors of bytes, or `Vec<u8>` in Rust and `Uint8Array` in JS
    -   `OptionBool`

## Package

Available on NPM:

```bash
npm i @scale-codec/core
```

## Core Concept

All codec functions in the library are built on top of these core types:

```ts
interface Walker {
    u8: Uint8Array
    idx: number
    view: DataView
}

interface Encode<T> {
    (value: T, walker: Walker): void
    sizeHint: (value: T) => number
}

type Decode<T> = (walker: Walker) => T
```

While **encoding**, firstly you should compute the value size prediction with a `.sizeHint(value)` function of `Encode` function, allocate buffer for it, put it into Walker interface and the pass it to the Encode function itself, which will mutate it's buffer and offset index. For example, here is a boolean encode function:

```ts
const encodeBool: Encode<boolean> = (value, walker) => {
    walker.u8[walker.idx++] = value ? 1 : 0
}

encodeBool.sizeHint = () => 1
```

While **decoding**, each decode function reading from Walker's buffer and incrementing it's offset index. For example, here is a boolean decode function:

```ts
const decodeBool: Decode<boolean> = (walker) => {
    return walker.u8[walker.idx++] === 1
}
```

Core library provides a very useful `WalkerImpl` utility to work with codecs:

```ts
const encoded: Uint8Array = WalkerImpl.encode(false, encodeBool)

const decoded: boolean = WalkerImpl.decode(encoded, decodeBool)
```

## Typing Tips

### Tuples

You can use TypeScript tuples to define tuple codecs safely:

```ts
import {
    createTupleEncoder,
    encodeU8,
    encodeStr,
} from '@scale-codec/core'

// (u8, str)
type MyTuple = [number, string]

const encoder = createTupleEncoder<MyTuple>([encodeU8, encodeStr])
```

### Structs

Struct codecs type definition is not completely automatic due to performance reasons and TypeScript limitations. Let's define a sample structure.

Definition in Rust:

```rust
use parity_scale_codec::{Encode, Decode};

#[derive(Encode, Decode)]
struct Message {
    author: String,
    timestamp: u128
}
```

Codecs with core library:

```ts
import {
    createStructEncoder,
    createStructDecoder,
    encodeU128,
    decodeU128,
    encodeStr,
    decodeStr,
} from '@scale-codec/core'

interface Message {
    author: string
    timestamp: bigint
}

const encoder = createStructEncoder<Message>([
    ['author', encodeStr],
    ['timestamp', encodeU128],
])

const decoder = createStructDecoder<Message>([
    ['author', decodeStr],
    ['timestamp', decodeU128],
])
```

Tuples passed to factories are typed only partially, i.e. TypeScript does not provide a guarantee that passed encoders/decoders schema will result into correct codecs. You should ensure by yourself that:

-   The order of fields is correct
-   All fields are defined

### Enums

::: tip
Enums handling is based on `@scale-codec/enum` package. More info at [Enums guide](./enum.md).
:::

::: tip
Core package re-exports everything from `@scale-codec/enum`, so:

```ts
import { Enum, Option /* ... */ } from '@scale-codec/core'
```

:::

Enums codecs _creation_ is even less strongly typed as struct ones. You are responsible to provide the correct schema. Let's define a sample enum.

In Rust:

```rust
enum Event {
    Focus,
    KeyPress(String)
}
```

In TypeScript:

```ts
import { Enum } from '@scale-codec/core'

type Event = Enum<'Focus' | ['KeyPress', string]>
```

Encoder:

```ts
import { createEnumEncoder, encodeStr } from '@scale-codec/core'

const encoder = createEnumEncoder<Event>({
    // map from <tag name> to <discriminant> for empty variants
    Focus: 0,
    // or [<discriminant>, <encode>] for valuable variants
    KeyPress: [1, encodeStr],
})
```

Decoder:

```ts
import { createEnumDecoder, decodeStr } from '@scale-codec/core'

const decoder = createEnumDecoder<Event>({
    // Map from <discriminant> to <tag name> for empty variants
    0: 'Focus',
    // or [<tag name>, <decode>] for valuable variants
    1: ['KeyPress', decodeStr],
})
```

Such approach is not very convenient to use, but it is worth it due to performance reasons.

Enums handling is more low-level then other tools because of typing issues. Thus, the responsibility to define encode & decode schema correctly is on the end user.

## Wrapping Up

This page doesn't describe all functions that it includes. For the more detailed API please refer to the related section.

Anyway, if you need to build a large namespace and avoid a lot of symmetric "encode/decode" functions, use tracking functionality for debugging with almost the same performance, you probably need to explore the next section.
