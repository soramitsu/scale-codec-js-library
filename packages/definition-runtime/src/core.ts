import { Decode, Encode, Walker, WalkerImpl, encodeFactory } from '@scale-codec/core'
import { trackDecode } from './tracking'

/**
 * General interface for any codec
 */
export interface Codec<Encoded, Decoded = Encoded> {
  encodeRaw: Encode<Encoded>
  decodeRaw: Decode<Decoded>
  fromBuffer: (src: ArrayBufferView) => Decoded
  toBuffer: (value: Encoded) => Uint8Array
  // name: () => string
}

export type CodecAny = Codec<any, any>

// /**
//  * General {@link Codec} implementation that includes fragments & tracking functionality
//  */
// export class CodecImpl<E, D = E> implements Codec<E, D> {
//     public encodeRaw: Encode<E>
//     public decodeRaw: Decode<D>
//     public name: () => string
//     public fromBuffer: (src: ArrayBufferView) => D
//     public toBuffer: (value: E) => Uint8Array

//     public constructor(name: string, encode: Encode<E>, decode: Decode<D>) {
//         this.encodeRaw = encode
//         this.decodeRaw = (walker) => trackDecode(name, walker, decode)
//         this.name = () => name
//         this.toBuffer = (value) => WalkerImpl.encode(value, this.encodeRaw)
//         this.fromBuffer = (src) => WalkerImpl.decode(src, this.decodeRaw)
//     }

//     // public fromBuffer(this: this, src: ArrayBufferView): D {
//     //     return WalkerImpl.decode(src, this.decodeRaw)
//     // }

//     // public toBuffer(this: this, value: E): Uint8Array {
//     //     return WalkerImpl.encode(value, this.encodeRaw)
//     // }

//     // public name(this: this): string {
//     //     return this._name
//     // }
// }

export function trackableCodec<E, D = E>(name: string, encode: Encode<E>, decode: Decode<D>): Codec<E, D> {
  const decodeTracked: Decode<D> = (walker) => trackDecode(name, walker, decode)

  return {
    // name: () => name,
    encodeRaw: encode,
    decodeRaw: decodeTracked,
    toBuffer: (value) => WalkerImpl.encode(value, encode),
    fromBuffer: (src) => WalkerImpl.decode(src, decodeTracked),
  }
}

export type CodecValueEncodable<T extends Codec<any>> = T extends Codec<infer E, any> ? E : never

export type CodecValueDecoded<T extends Codec<any>> = T extends Codec<any, infer D> ? D : never

// /**
//  * Special {@link Codec} implementation to that wraps another codec and dispatches it via its getter.
//  * With this utility it is easy to implement cyclic dependencies between codecs.
//  *
//  * @remarks
//  *
//  * TODO optimize and cache got codec after the first dispatch?
//  */
// export class DynCodec<C extends CodecAny> implements Codec<CodecValueEncodable<C>, CodecValueDecoded<C>> {
//     public encodeRaw: Encode<CodecValueEncodable<C>>

//     public decodeRaw: Decode<CodecValueDecoded<C>>

//     public readonly codecGetter: () => C

//     public constructor(getter: () => C) {
//         this.codecGetter = getter

//         this.encodeRaw = encodeFactory(
//             (val, walker) => getter().encodeRaw(val, walker),
//             (val) => getter().encodeRaw.sizeHint(val),
//         )

//         this.decodeRaw = (walker) => getter().decodeRaw(walker)
//     }

//     public fromBuffer(this: this, src: ArrayBufferView): CodecValueDecoded<C> {
//         return this.codecGetter().fromBuffer(src)
//     }

//     public toBuffer(this: this, value: CodecValueEncodable<C>): Uint8Array {
//         return this.codecGetter().toBuffer(value)
//     }

//     public name(this: this): string {
//         return this.codecGetter().name()
//     }
// }

// /**
//  * See {@link DynCodec}. This function is a shorthand factory that is more minimize-friendly than `new DynCodec(...)`
//  */
// export function dynCodec<C extends CodecAny>(getter: () => C): DynCodec<C> {
//     return new DynCodec(getter)
// }

export function dynCodec<C extends CodecAny>(getter: () => C): Codec<CodecValueEncodable<C>, CodecValueDecoded<C>> {
  let codec: C | undefined

  const getCodec = (): C => {
    if (!codec) {
      codec = getter()
    }
    return codec
  }

  return {
    // name: () => (codec ?? getCodec()).name(),
    encodeRaw: encodeFactory(
      (val, walker) => (codec ?? getCodec()).encodeRaw(val, walker),
      (val) => (codec ?? getCodec()).encodeRaw.sizeHint(val),
    ),
    decodeRaw: (walker) => (codec ?? getCodec()).decodeRaw(walker),
    toBuffer: (value) => (codec ?? getCodec()).toBuffer(value),
    fromBuffer: (src) => (codec ?? getCodec()).fromBuffer(src),
  }
}
