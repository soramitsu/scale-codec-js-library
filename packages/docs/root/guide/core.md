# Core package

`@scale-codec/core` package contains all necessary low-level tools to perform encoding and decoding between SCALE binaries and their JS representations.

## Contents

[[toc]]

## Supported types

-   Primitives
    -   Integers - signed & unsigned, 8/16/32/64/128 bits, as [`JSBI` (`BigInt` replacement for compatibility)](https://github.com/GoogleChromeLabs/jsbi), and...
        -   Compact (special integers representation) - as `JSBI` too
    -   String - as native JS `String`
    -   Boolean - as native JS `Boolean`
    -   Void (just for consistence) - as `null`
-   Complex
    -   Fixed-length Array - as JS `Array`, and...
        -   `[u8; len]` as JS `Uint8Array`
    -   Vector - as JS `Array`, and...
        -   `Vec<u8>` as JS `Uint8Array`
    -   Set - as native JS `Set`
    -   Map - as native JS `Map`
    -   Struct - as JS object
    -   Tuple - as JS `Array`
    -   Enum (+ special codec for OptionBool) - as `Enum` from [`@scale-codec/enum` package](./enum)

## Usage examples

<script setup>
import BigIntPlayground from './components/BigIntPlayground.vue'
import BigIntEncode from './components/BigIntEncode.vue'
import CompactPlayground from './components/CompactPlayground.vue'
import CompactEncode from './components/CompactEncode.vue'
import StrEncode from './components/StrEncode.vue'
import CoreStructResult from './components/CoreStructResult.vue'
</script>

### Integers

```ts
import { encodeBigInt, JSBI } from '@scale-codec/core';

const num = JSBI.BigInt(5_012_009);
const encoded = encodeBigInt(num, {
    bits: 32,
    signed: false,
    endianness: 'le',
});

console.log(encoded);
```

Result:

<BigIntEncode :bits="32" endianness="le" num="5012009" />

[Playground](#bigint-playground)

### Compact

```ts
import { encodeCompact, JSBI } from '@scale-codec/core';
import { hexifyBytes } from '@scale-codec/util';

for (const num of [1, 34_000, 891_000_000, '24141828384918234']) {
    const encoded = encodeCompact(JSBI.BigInt(num));
    console.log('%s: %s', num, hexifyBytes(encoded));
}
```

Result:

<p><template v-for="x in [1, 34_000, 891_000_000, '24141828384918234']">{{ x }}: <code><CompactEncode :num="String(x)" hex /></code><br></template></p>

[Playground](#compact-playground)

### Strings

```ts
import { encodeStr } from '@scale-codec/core';
import { hexifyBytes } from '@scale-codec/util';

console.log(hexifyBytes(encodeStr('Лев Николаевич Толстой')));
```

Result:

<code><StrEncode val="Лев Николаевич Толстой" /></code>

### Other primitives

```ts
import { encodeBool, decodeBool, encodeVoid, decodeVoid } from '@scale-codec/core';

function assertEq(a: any, b: any) {
    // asserting deep equality between a & b...
}

assertEq(encodeBool(false), new Uint8Array([0]));
assertEq(encodeBool(true), new Uint8Array([1]));
assertEq(decodeBool(new Uint8Array([1])), [true, 1]);

assertEq(decodeVoid(new Uint8Array()), [null, 0]);
assertEq(decodeVoid(new Uint8Array([1, 2, 4, 5])), [null, 0]);
```

### Struct

Definition in Rust:

```rust
use parity_scale_codec::{Encode, Decode};

#[derive(Encode, Decode)]
struct Message {
    author: String,
    timestamp: u128
}
```

Definition & encoding in JavaScript:

```ts
import { encodeStruct, JSBI, encodeStrCompact, encodeBigInt } from '@scale-codec/core';
import { hexifyBytes } from '@scale-codec/util';

interface Message {
    author: string;
    timestamp: JSBI;
}

const msg: Message = {
    author: 'Clara',
    timestamp: JSBI.BigInt('16488182899412'),
};

const msgEncoded = encodeStruct(
    msg,
    {
        author: encodeStrCompact,
        timestamp: (v) =>
            encodeBigInt(v, {
                bits: 128,
                signed: false,
                endianness: 'le',
            }),
    },
    ['author', 'timestamp'],
);

console.log(hexifyBytes(msgEncoded));
```

Output:

<code><CoreStructResult /></code>

### Enum

::: tip
Enums handling is based on `@scale-codec/enum` package. More info at [Enums guide](./enum.md).
:::

Enums handling is more low-level then other tools because of typing issues. Thus, the responsibility to define encode & decode schema correctly is on the end user.

```ts
import { Enum, Valuable, encodeEnum, decodeEnum, encodeStr, decodeStr } from '@scale-codec/core';

type Event = Enum<{
    Focus: null;
    KeyPress: Valuable<string>;
}>;

const myEvent = decodeEnum<Event>(new Uint8Array(/* ... */), {
    // map from discriminants to variant names
    // + optional decode fn for valuable variants
    0: { v: 'Focus' },
    1: { v: 'KeyPress', decode: decodeStr },
});

const encoded = encodeEnum(myEvent, {
    // map from variant names to discriminants
    // + optional encode fn for valuable variants
    Focus: { d: 0 },
    KeyPress: { d: 1, encode: encodeStr },
});
```

### Array, Vec, Tuple

```ts
import { decodeTuple, decodeArray, decodeVec, decodeStr, decodeBool } from '@scale-codec/core';

// [String; 10]
const strings = decodeArray(new Uint8Array(/* ... */), decodeStr, 10);

// (String, bool)
const [str, bool] = decodeTuple(new Uint8Array(/* ... */), [decodeStr, decodeBool]);

// Vec<bool>
const booleans = decodeVec(new Uint8Array(/* ... */), decodeBool);
```

### Map & Set

```ts
import { encodeSet, encodeMap, encodeStr, encodeBool } from '@scale-codec/core';

encodeMap(
    new Map<string, boolean>([
        ['foo', false],
        ['bar', true],
    ]),
    encodeStr,
    encodeBool,
);

encodeSet(new Set<string>(['a', 'b', 'c']), encodeStr);
```

## Play

### BigInt Playground

<BigIntPlayground class="mt-4" />

### Compact Playground

<CompactPlayground class="mt-4" />

## Also

-   [API](../api/core)

### Why `JSBI` for numbers?

Because it can be easily migrated to the native BigInt in the future with a [special JSBI's Babel plugin](https://github.com/GoogleChromeLabs/babel-plugin-transform-jsbi-to-bigint).
