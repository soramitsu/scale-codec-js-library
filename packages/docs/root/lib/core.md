# Core package

`@scale-codec/core` package contains all necessary low-level tools to perform encoding and decoding between SCALE binaries and their JS representations.

## Contents

[[toc]]

## Supported types

-   Primitives
    -   Integers - signed & unsigned, 8/16/32/64/128 bits, as [`JSBI` (`BigInt` replacement for compatibility)](https://github.com/GoogleChromeLabs/jsbi), and...
        -   Compact (special integers representation) - as `JSBI` too
    -   String - as native JS `String`
    -   Boolean - as native JS `Boolean`
    -   Void (just for consistence) - as `null`
-   Complex
    -   Fixed-length Array - as JS `Array`, and...
        -   `[u8; len]` as JS `Uint8Array`
    -   Vector - as JS `Array`, and...
        -   `Vec<u8>` as JS `Uint8Array`
    -   Set - as native JS `Set`
    -   Map - as native JS `Map`
    -   Struct - as JS object
    -   Tuple - as JS `Array`
    -   Enum (+ special codec for OptionBool) - as `Enum` from [`@scale-codec/enum` package](./enum)

## Usage examples

<script setup>
import BigIntPlayground from './components/BigIntPlayground.vue'
import BigIntEncode from './components/BigIntEncode.vue'
import CompactPlayground from './components/CompactPlayground.vue'
import CompactEncode from './components/CompactEncode.vue'
import StrEncode from './components/StrEncode.vue'
import CoreStructResult from './components/CoreStructResult.vue'
</script>

### Integers

```ts
import { encodeBigInt, JSBI } from '@scale-codec/core';

const num = JSBI.BigInt(5_012_009);
const encoded = encodeBigInt(num, {
    bits: 32,
    signed: false,
    endianness: 'le',
});

console.log(encoded);
```

Result:

<BigIntEncode :bits="32" endianness="le" num="5012009" />

[Playground](#bigint-playground)

### Compact

```ts
import { encodeCompact, JSBI } from '@scale-codec/core';
import { hexifyBytes } from '@scale-codec/util';

for (const num of [1, 34_000, 891_000_000, '24141828384918234']) {
    const encoded = encodeCompact(JSBI.BigInt(num));
    console.log('%s: %s', num, hexifyBytes(encoded));
}
```

Result:

<div><template v-for="x in [1, 34_000, 891_000_000, '24141828384918234']">{{ x }}: <CompactEncode :num="String(x)" hex /><br></template></div>

[Playground](#compact-playground)

### Strings

```ts
import { encodeStrCompact } from '@scale-codec/core';
import { hexifyBytes } from '@scale-codec/util';

console.log(hexifyBytes(encodeStrCompact('Лев Николаевич Толстой')));
```

Result:

<StrEncode val="Лев Николаевич Толстой" />

### Other primitives

```ts
import { encodeBool, decodeBool, encodeVoid, decodeVoid } from '@scale-codec/core';

function assertEq(a: any, b: any) {
    // asserting deep equality between a & b...
}

assertEq(encodeBool(false), new Uint8Array([0]));
assertEq(encodeBool(true), new Uint8Array([1]));
assertEq(decodeBool(new Uint8Array([1])), [true, 1]);

assertEq(decodeVoid(new Uint8Array()), [null, 0]);
assertEq(decodeVoid(new Uint8Array([1, 2, 4, 5])), [null, 0]);
```

### Struct

Definition in Rust:

```rust
use parity_scale_codec::{Encode, Decode};

#[derive(Encode, Decode)]
struct Message {
    author: String,
    timestamp: u128
}
```

Definition & encoding in JavaScript:

```ts
import { encodeStruct, JSBI, encodeStrCompact, encodeBigInt } from '@scale-codec/core';
import { hexifyBytes } from '@scale-codec/util';

interface Message {
    author: string;
    timestamp: JSBI;
}

const msg: Message = {
    author: 'Clara',
    timestamp: JSBI.BigInt('16488182899412'),
};

const msgEncoded = encodeStruct(
    msg,
    {
        author: encodeStrCompact,
        timestamp: (v) =>
            encodeBigInt(v, {
                bits: 128,
                signed: false,
                endianness: 'le',
            }),
    },
    ['author', 'timestamp'],
);

console.log(hexifyBytes(msgEncoded));
```

Output:

<CoreStructResult />

### More example

Todo for:

-   Map
-   Set
-   Tuple
-   Array & Vec
-   Enum

## Play

### BigInt Playground

<BigIntPlayground class="mt-4" />

### Compact Playground

<CompactPlayground class="mt-4" />

## Also

-   [API](../api/core)
