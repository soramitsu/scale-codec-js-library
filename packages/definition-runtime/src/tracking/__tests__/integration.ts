import { Decode, Walker, WalkerImpl } from '@scale-codec/core'
import { Enum } from '../../lib'
import { setCurrentTracker, DecodeTraceCollector, buildDecodeTraceStepsFmt, CodecTracker, DecodeTrace } from '../index'
// useful for tests here too
import { StructA } from '@scale-codec/definition-compiler/tests/samples/unwrapCheck'

describe('Collecting big decode trace and formatting it', () => {
    class TestTracker implements CodecTracker {
        public lastTrace: null | DecodeTrace = null
        public lastWalker: null | Walker = null
        private errored = false
        private depth = 0
        private tracer = new DecodeTraceCollector()

        public decode<T>(loc: string, walker: Walker, decode: Decode<T>): T {
            try {
                this.lastWalker = walker
                this.depth++
                this.tracer.decodeStart(loc, walker)
                const result = decode(walker)
                const trace = this.tracer.decodeSuccess(walker, result)
                trace && (this.lastTrace = trace)
                return result
            } catch (err) {
                if (!this.errored) {
                    this.errored = true
                    this.lastTrace = this.tracer.decodeError(err)
                }
                throw err
            } finally {
                if (!--this.depth) {
                    this.errored = false
                }
            }
        }

        public refineDecodeLoc<T>(loc: string, decode: () => T) {
            this.tracer.refineLoc(loc)
            return decode()
        }
    }

    const fragment = StructA.wrap({
        primitive: true,
        enum: Enum.variant('Empty'),
        map: new Map([
            ['test str', ['tuple value']],
            ['another key', ['tuple value']],
        ]),
        set: new Set([['tuple value'], ['another tuple']]),
        tuple: ['tuple value'],
        array: [true, true, true],
        bytesArray: new Uint8Array([8, 1, 2, 3, 3]),
        vec: [Enum.variant('Opt', Enum.variant('None')), Enum.variant('Res', Enum.variant('Err', 'test str'))],
        alias: ['test str'],
    })

    afterEach(() => {
        setCurrentTracker(null)
    })

    test('Success case', () => {
        const tracker = new TestTracker()
        setCurrentTracker(tracker)

        // eslint-disable-next-line @typescript-eslint/no-unused-expressions
        StructA.fromBuffer(fragment.bytes).value

        expect(tracker.lastTrace).toBeTruthy()
        expect(buildDecodeTraceStepsFmt(tracker.lastTrace!, tracker.lastWalker!).assemble()).toMatchSnapshot()
    })

    test('Error case', () => {
        const tracker = new TestTracker()
        setCurrentTracker(tracker)

        const bytes = fragment.bytes
        const copy = new Uint8Array([...bytes]).fill(255, 20, 30)
        expect(() => StructA.fromBuffer(copy).value).toThrow()

        expect(tracker.lastTrace).toBeTruthy()
        expect(buildDecodeTraceStepsFmt(tracker.lastTrace!, tracker.lastWalker!).assemble()).toMatchSnapshot()
    })
})
