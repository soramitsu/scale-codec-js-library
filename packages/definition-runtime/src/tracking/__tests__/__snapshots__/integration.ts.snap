// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collecting big decode trace and formatting it Error case 1`] = `
Array [
  "<[01] 20 74 65 73 74 20 73 74 72 02 08 20 74 65 73 74 20 73 74... (len: 118, used: 1)>
  StructA ğŸ¡ª .primitive ğŸ¡ª Bool: %O
<[20 74 65 73 74 20 73 74 72] 02 08 20 74 65 73 74 20 73 74 ff... (len: 117, used: 9)>
  StructA ğŸ¡ª .alias ğŸ¡ª TupleA ğŸ¡ª .0 ğŸ¡ª Str: %O
  StructA ğŸ¡ª .alias ğŸ¡ª TupleA: %O
<[02] 08 20 74 65 73 74 20 73 74 ff ff ff ff ff ff ff ff ff ff... (len: 108, used: 1)>
  StructA ğŸ¡ª .enum ğŸ¡ª EnumA: %O
<20 74 65 73 74 20 73 74 ff ff ff ff ff ff ff ff ff ff 6c 75... (len: 106)>
  StructA ğŸ¡ª .map ğŸ¡ª MapA ğŸ¡ª <key> ğŸ¡ª Str: ERROR - %s
  StructA ğŸ¡ª .map ğŸ¡ª MapA: <not computed>
  StructA: <not computed>
",
  true,
  "test str",
  Array [
    "test str",
  ],
  Object {
    "tag": "Empty",
  },
  [TypeError: The encoded data was not valid for encoding utf-8],
]
`;

exports[`Collecting big decode trace and formatting it Success case 1`] = `
Array [
  "<[01] 20 74 65 73 74 20 73 74 72 02 08 20 74 65 73 74 20 73 74... (len: 118, used: 1)>
  StructA ğŸ¡ª .primitive ğŸ¡ª Bool: %O
<[20 74 65 73 74 20 73 74 72] 02 08 20 74 65 73 74 20 73 74 72... (len: 117, used: 9)>
  StructA ğŸ¡ª .alias ğŸ¡ª TupleA ğŸ¡ª .0 ğŸ¡ª Str: %O
  StructA ğŸ¡ª .alias ğŸ¡ª TupleA: %O
<[02] 08 20 74 65 73 74 20 73 74 72 2c 74 75 70 6c 65 20 76 61... (len: 108, used: 1)>
  StructA ğŸ¡ª .enum ğŸ¡ª EnumA: %O
<[20 74 65 73 74 20 73 74 72] 2c 74 75 70 6c 65 20 76 61 6c 75... (len: 106, used: 9)>
  StructA ğŸ¡ª .map ğŸ¡ª MapA ğŸ¡ª <key> ğŸ¡ª Str: %O
<[2c 74 75 70 6c 65 20 76 61 6c 75 65] 2c 61 6e 6f 74 68 65 72... (len: 97, used: 12)>
  StructA ğŸ¡ª .map ğŸ¡ª MapA ğŸ¡ª <value> ğŸ¡ª TupleA ğŸ¡ª .0 ğŸ¡ª Str: %O
  StructA ğŸ¡ª .map ğŸ¡ª MapA ğŸ¡ª <value> ğŸ¡ª TupleA: %O
<[2c 61 6e 6f 74 68 65 72 20 6b 65 79] 2c 74 75 70 6c 65 20 76... (len: 85, used: 12)>
  StructA ğŸ¡ª .map ğŸ¡ª MapA ğŸ¡ª <key> ğŸ¡ª Str: %O
<[2c 74 75 70 6c 65 20 76 61 6c 75 65] 08 2c 74 75 70 6c 65 20... (len: 73, used: 12)>
  StructA ğŸ¡ª .map ğŸ¡ª MapA ğŸ¡ª <value> ğŸ¡ª TupleA ğŸ¡ª .0 ğŸ¡ª Str: %O
  StructA ğŸ¡ª .map ğŸ¡ª MapA ğŸ¡ª <value> ğŸ¡ª TupleA: %O
  StructA ğŸ¡ª .map ğŸ¡ª MapA: %O
<[2c 74 75 70 6c 65 20 76 61 6c 75 65] 34 61 6e 6f 74 68 65 72... (len: 60, used: 12)>
  StructA ğŸ¡ª .set ğŸ¡ª SetA ğŸ¡ª TupleA ğŸ¡ª .0 ğŸ¡ª Str: %O
  StructA ğŸ¡ª .set ğŸ¡ª SetA ğŸ¡ª TupleA: %O
<[34 61 6e 6f 74 68 65 72 20 74 75 70 6c 65] 01 01 01 08 01 02... (len: 48, used: 14)>
  StructA ğŸ¡ª .set ğŸ¡ª SetA ğŸ¡ª TupleA ğŸ¡ª .0 ğŸ¡ª Str: %O
  StructA ğŸ¡ª .set ğŸ¡ª SetA ğŸ¡ª TupleA: %O
  StructA ğŸ¡ª .set ğŸ¡ª SetA: %O
<[01] 01 01 08 01 02 03 03 08 00 00 01 01 20 74 65 73 74 20 73... (len: 34, used: 1)>
  StructA ğŸ¡ª .array ğŸ¡ª ArrayA ğŸ¡ª Bool: %O
<[01] 01 08 01 02 03 03 08 00 00 01 01 20 74 65 73 74 20 73 74... (len: 33, used: 1)>
  StructA ğŸ¡ª .array ğŸ¡ª ArrayA ğŸ¡ª Bool: %O
<[01] 08 01 02 03 03 08 00 00 01 01 20 74 65 73 74 20 73 74 72... (len: 32, used: 1)>
  StructA ğŸ¡ª .array ğŸ¡ª ArrayA ğŸ¡ª Bool: %O
  StructA ğŸ¡ª .array ğŸ¡ª ArrayA: %O
<[08 01 02 03 03] 08 00 00 01 01 20 74 65 73 74 20 73 74 72 2c... (len: 31, used: 5)>
  StructA ğŸ¡ª .bytesArray ğŸ¡ª BytesArrayA: %O
<[00] 01 01 20 74 65 73 74 20 73 74 72 2c 74 75 70 6c 65 20 76... (len: 24, used: 1)>
  StructA ğŸ¡ª .vec ğŸ¡ª VecEnumA ğŸ¡ª EnumA ğŸ¡ª ::Opt ğŸ¡ª OptionA: %O
  StructA ğŸ¡ª .vec ğŸ¡ª VecEnumA ğŸ¡ª EnumA: %O
<[20 74 65 73 74 20 73 74 72] 2c 74 75 70 6c 65 20 76 61 6c 75... (len: 21, used: 9)>
  StructA ğŸ¡ª .vec ğŸ¡ª VecEnumA ğŸ¡ª EnumA ğŸ¡ª ::Res ğŸ¡ª ResultA ğŸ¡ª ::Err ğŸ¡ª Str: %O
  StructA ğŸ¡ª .vec ğŸ¡ª VecEnumA ğŸ¡ª EnumA ğŸ¡ª ::Res ğŸ¡ª ResultA: %O
  StructA ğŸ¡ª .vec ğŸ¡ª VecEnumA ğŸ¡ª EnumA: %O
  StructA ğŸ¡ª .vec ğŸ¡ª VecEnumA: %O
<[2c 74 75 70 6c 65 20 76 61 6c 75 65] (len = used = 12)>
  StructA ğŸ¡ª .tuple ğŸ¡ª TupleA ğŸ¡ª .0 ğŸ¡ª Str: %O
  StructA ğŸ¡ª .tuple ğŸ¡ª TupleA: %O
  StructA: %O
",
  true,
  "test str",
  Array [
    "test str",
  ],
  Object {
    "tag": "Empty",
  },
  "test str",
  "tuple value",
  Array [
    "tuple value",
  ],
  "another key",
  "tuple value",
  Array [
    "tuple value",
  ],
  Map {
    "test str" => Array [
      "tuple value",
    ],
    "another key" => Array [
      "tuple value",
    ],
  },
  "tuple value",
  Array [
    "tuple value",
  ],
  "another tuple",
  Array [
    "another tuple",
  ],
  Set {
    Array [
      "tuple value",
    ],
    Array [
      "another tuple",
    ],
  },
  true,
  true,
  true,
  Array [
    true,
    true,
    true,
  ],
  Uint8Array [
    8,
    1,
    2,
    3,
    3,
  ],
  Object {
    "tag": "None",
  },
  Object {
    "tag": "Opt",
    "value": Object {
      "tag": "None",
    },
  },
  "test str",
  Object {
    "tag": "Err",
    "value": "test str",
  },
  Object {
    "tag": "Res",
    "value": Object {
      "tag": "Err",
      "value": "test str",
    },
  },
  Array [
    Object {
      "tag": "Opt",
      "value": Object {
        "tag": "None",
      },
    },
    Object {
      "tag": "Res",
      "value": Object {
        "tag": "Err",
        "value": "test str",
      },
    },
  ],
  "tuple value",
  Array [
    "tuple value",
  ],
  Object {
    "alias": Array [
      "test str",
    ],
    "array": Array [
      true,
      true,
      true,
    ],
    "bytesArray": Uint8Array [
      8,
      1,
      2,
      3,
      3,
    ],
    "enum": Object {
      "tag": "Empty",
    },
    "map": Map {
      "test str" => Array [
        "tuple value",
      ],
      "another key" => Array [
        "tuple value",
      ],
    },
    "primitive": true,
    "set": Set {
      Array [
        "tuple value",
      ],
      Array [
        "another tuple",
      ],
    },
    "tuple": Array [
      "tuple value",
    ],
    "vec": Array [
      Object {
        "tag": "Opt",
        "value": Object {
          "tag": "None",
        },
      },
      Object {
        "tag": "Res",
        "value": Object {
          "tag": "Err",
          "value": "test str",
        },
      },
    ],
  },
]
`;
